src/qofono.cpp

```diff
@@ -403,8 +1928,24 @@ QOfono::QOfono()
 
 QString QOfono::mobileCountryCodeToAlpha2CountryCode(int mcc)
 {
-    if (mcc < 0 || mcc >= COUNTRY_CODES_LENGTH) {
-        qWarning() << Q_FUNC_INFO << "given unknown Mobile Country Code:" << mcc;
+    int low = 0;
+    int high = (sizeof(qofonoMccList)/sizeof(qofonoMccList[0]))-1;
+
+    while (low <= high) {
+        int mid = (low + high)/2;
+        int val = qofonoMccList[mid].mcc;
+        if (val < mcc) {
+            low = mid + 1;
+        } else if (val > mcc) {
+            high = mid - 1;
+        } else {
+            /* MCC is found. Back up a litte to the first entry */
+            while (mid > 0 && qofonoMccList[mid-1].mcc == val) mid--;
+            return QString(qofonoMccList[mid].cc);
```

Comment:
With the current table this algorithm has pretty flat performance (10 or 11 steps, mostly 11) while  binary search + backup takes anywhere from 3 (545/KI) or 4 (293/SI) up to 2+48=50 (310/US), 3+76=79 (440/JP) or 3+83=86 (405/IN), summing up both binary and linear search steps. It definitely makes more sense to optimize for US, Japan or India as opposed to Kiribati or even Slovenia (even though currently we might have more customers in Slovenia than in Japan). Also, it has one comparison per binary search loop while the original one had two.

Your suggestion clearly wins. Thanks.